connectors/sources/sharepoint.py

```diff
@@ -0,0 +1,715 @@
+            return
+        await self.session.close()
+
+    async def _api_call(self, url_name, url="", **url_kwargs):
+        """Make an API call to the Sharepoint Server/Online
+
+        Args:
+            url_name (str): Sharepoint url name to be executed.
+            url(str, optional): Paginated url for drive and list items. Defaults to "".
+            url_kwargs (dict): Url kwargs to format the query.
+        Raises:
+            exception: An instance of an exception class.
+
+        Yields:
+            data: API response.
+        """
+        retry = 0
+        # If pagination happens for list and drive items then next pagination url comes in response which will be passed in url field.
+        if url == "":
+            url = URLS[url_name].format(**url_kwargs)
+        while True:
+            try:
+                async with self.session.get(
+                    url=url,
+                    ssl=self.ssl_ctx,
+                    headers={"Authorization": f"Bearer {self.access_token}"}
+                    if self.is_cloud
+                    else None,
+                ) as result:
+                    if url_name == ATTACHMENT:
+                        yield result
+                    else:
+                        yield await result.json()
+                    break
+            except Exception as exception:
+                if isinstance(
+                    exception, aiohttp.client_exceptions.ClientResponseError
+                ) and "token has expired" in exception.headers.get(
+                    "x-ms-diagnostics", ""
+                ):
+                    await self._set_access_token()
+                elif isinstance(
+                    exception, aiohttp.client_exceptions.ServerDisconnectedError
+                ):
+                    await self.close()
+                    self.session = await self._generate_session()
+                retry += 1
+                if retry > self.retry_count:
+                    raise exception
+                logger.warning(
+                    f"Retry count: {retry} out of {self.retry_count}. Exception: {exception}"
+                )
+                await asyncio.sleep(RETRY_INTERVAL**retry)
+
+    def format_document(
+        self,
+        item,
+        document_type,
+        item_type=None,
+        file_name=None,
+    ):
+        """Prepare key mappings for sites, lists, list items and drive items
+
+        Args:
+            item (dictionary): Document from Sharepoint.
+            document_type(string): Type of document(i.e. list_item and drive_item).
+            item_type(string, optional): Type of item i.e. File or Folder. Defaults to None.
+            file_name(string, optional): Name of file. Defaults to None.
+
+        Returns:
+            dictionary: Modified document with the help of adapter schema.
+        """
+        document = {"type": document_type}
+
+        if document_type in [LISTS, DOCUMENT_LIBRARY]:
```

Comment:
Consider using the match and case statement: https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching

Reply:
if we will go with match case then we can't use constant. We will have to use hard coded values
```python
match document_type:
    case "lists" | "document_library":
        document["url"] = urljoin(
            self.host_url, item["RootFolder"]["ServerRelativeUrl"]
        )
    case "drive_item":
        document.update(
            {
                "_id": item["GUID"],
                "size": item.get("File", {}).get("Length"),
                "url": urljoin(self.host_url, item[item_type]["ServerRelativeUrl"]),
                "type": item_type,
            }
        )
    case "list_item":
        document.update(
            {
                "_id": item["_id"] if "_id" in item.keys() else item["GUID"],
                "file_name": file_name,
                "size": item.get("size"),
                "url": item["url"],
            }
                )
```